blog/2023-11-20-why-is-windmill-the-fastest-workflow-engine/index.mdx

```diff
@@ -0,0 +1,174 @@
+different resources depending on the roles of the caller.
+
+## So why is windmill very fast
+
+In a workflow engine, the efficiency depend:
+
+- efficiency to compute transitions, the new jobs to schedule given the last job that completed, and the efficiency of the workers themselves to pull scheduled jobs and run them
+- efficiency to pass data between steps
+- efficiency of the worker to pull job, start executing the code, and then submit result and new sate
+
+Windmill is extremely fast because on all 3 aspects, it uses a simple design, optimized everywhere, that leverage maximally both Postgresql and Rust.
+
+## System design and Queue
+
+Windmill offers a single binary (compiled from Rust) than can run both as the api server, the worker or both.
+Both workers and servers are connected to Postgresql but not to each other.
+The servers only expose the api and the frontend. The queue is implemented in PG itself.
+Jobs can be triggered externally by calling the api which will push new jobs to the queue.
+
+Jobs are stored in 2 tables in PG:
+
+- queue (while the job is not completed, even when running)
+- completed_job
+
+Jobs are not removed from the queue upon start, but their field `running` is set to true. Queue is implemented with the traditional `UPDATE SKIP LOCKED`
+
+```sql
+UPDATE queue
+SET running = true
+, started_at = coalesce(started_at, now())
+, last_ping = now()
+, suspend_until = null
+WHERE id = (
+    SELECT id
+    FROM queue
+    WHERE running = false AND scheduled_for <= now() AND tag = ANY($1)
+    ORDER BY priority DESC NULLS LAST, scheduled_for, created_at
+    FOR UPDATE SKIP LOCKED
+    LIMIT 1
+)
+RETURNING *
+```
+
+This is as fast as you can get with Postgresql, as long as every field is properly indexed. Some variation of this delete the job from the queue instead of updating a flag, or set a time at which it is invisible for others to be pulled. It's all the same principle under the hood.
+
+When a flow job is pushed, its inputs, pointers to an immutable flow definition is pushed alongside the job and initial flow state (see under) is pushed within the job (it's a big row!)
+
+A worker will then pick the flow job, read the flow definition and flow state, realize that it needs to push the first step in the queue and that's it. This is the initial transition of the flow.
+
+The workers pull jobs one at a time, run them to completion, and progres the state if the job is part of a flow job (which is a separate job). The servers themselves do no orchestration, every flow progress is run by the worker themselves.
+
+## States
+
+Workflow engines represents jobs as a finite state machine [(FSM)](https://en.wikipedia.org/wiki/Finite-state_machine). The 4 main states commonly used are:
+
+- Waiting for pre-requisites (right events like a webhook to have been received, or all the depenency jobs to have been completed)
```

Comment:
(events like a webhook to be received, or all the dependency jobs to be completed)